File: 0079-word-search/0079-word-search.js
/**
 * @param {character[][]} board
 * @param {string} word
 * @return {boolean}
 */
var exist = function(board, word) {
    const rows = board.length
    const cols = board[0].length

    function helper(r, c, i){
        if(i === word.length){
            return true
        }
        if(r < 0 || c < 0 
        || r >= rows || c >= cols 
        || board[r][c] !== word[i] 
        || !board[r][c]){
            return false
        }

        const temp = board[r][c]
        board[r][c] = null
        const res = helper(r + 1, c, i + 1) 
                    || helper(r - 1, c, i + 1) 
                    || helper(r, c + 1, i + 1) 
                    || helper(r, c - 1, i + 1)
        board[r][c] = temp
        return res
    }

    for(let r = 0; r < rows; r++){
        for(let c = 0; c < cols; c++){
            if(board[r][c] === word[0]){
                if(helper(r, c, 0)){
                    return true
                }
            }
        }
    }
    return false
};

File: 0079-word-search/0079-word-search.py
class Solution:
    def exist(self, board: List[List[str]], word: str) -> bool:
        rows = len(board)
        cols = len(board[0])

        for i in range(rows):
            for j in range(cols):
                if board[i][j] == word[0]:
                    if self.find_string(i, j, board, word, 0):
                        return True
        
        return False
    
    def find_string(self, row, col, board, word, i):
        if i == len(word):
            return True
        
        if (row < 0 or row >= len(board) 
        or col < 0 or col >= len(board[0]) 
        or board[row][col] != word[i]):
            return False

        temp = board[row][col]
        board[row][col] = "#"

        found = (
            self.find_string(row+1, col, board, word, i+1) or
            self.find_string(row-1, col, board, word, i+1) or
            self.find_string(row, col+1, board, word, i+1) or
            self.find_string(row, col-1, board, word, i+1)
        )

        board[row][col] = temp
        
        return found